nixos/modules/security/apparmor.nix

```diff
@@ -1,59 +1,190 @@
+    '' + head (match "^.*\\[qualifiers](.*)" # Drop the original [settings] section.
+                     (readFile "${pkgs.apparmor-utils}/etc/apparmor/logprof.conf"));
+
+    boot.kernelParams = [ "apparmor=1" "security=apparmor" ];
+
+    systemd.services.apparmor = {
+      after = [
+        "local-fs.target"
+        "systemd-journald-audit.socket"
+      ];
+      before = [ "sysinit.target" ];
+      wantedBy = [ "multi-user.target" ];
+      restartTriggers = [
+        etc."apparmor/parser.conf".source
+        etc."apparmor.d".source
+      ];
+      unitConfig = {
+        Description="Load AppArmor policies";
+        DefaultDependencies = "no";
+        ConditionSecurity = "apparmor";
+      };
+      # Reloading instead of restarting enables to load new AppArmor profiles
+      # without necessarily restarting all services which have Requires=apparmor.service
+      # It works by:
+      # - Adding or replacing into the kernel profiles enabled in cfg.policies
+      #   (because AppArmor can do that without stopping the processes already confined).
+      # - Removing from the kernel any profile whose name is not
+      #   one of the names within the content of the profiles in cfg.policies.
+      # - Killing the processes which are unconfined but now have a profile loaded
+      #   (because AppArmor can only confine new processes).
+      reloadIfChanged = true;
+      # Avoid searchs in /usr/share/locale/
+      environment.LANG="C";
+      serviceConfig = let
+        enabledPolicies = lib.attrValues (lib.filterAttrs (n: p: p.enable) cfg.policies);
+        removeDisabledProfiles = pkgs.writeShellScript "apparmor-remove" ''
+          set -eux
+
+          loadedProfiles=$(mktemp)
+          enabledProfiles=$(mktemp)
+          trap "rm -f $loadedProfiles $enabledProfiles" EXIT
+
+          sed -e "s/ (\(enforce\|complain\))$//" /sys/kernel/security/apparmor/profiles |
+          sort -u >"$loadedProfiles"
+
+          ${pkgs.apparmor-parser}/bin/apparmor_parser --names /dev/null ${
+            lib.concatMapStrings (p: "\\\n "+p.profile) enabledPolicies} |
+          sort -u >"$enabledProfiles"
+
+          comm -23 "$loadedProfiles" "$enabledProfiles" |
+          while IFS=$'\n\r' read -r profile
+          do printf %s "$profile" >/sys/kernel/security/apparmor/.remove
+          done
+        '';
+        killUnconfinedConfinables = pkgs.writeShellScript "apparmor-kill" ''
+          set -eux
+          ${pkgs.apparmor-utils}/bin/aa-status --json |
+          ${pkgs.jq}/bin/jq --raw-output '.processes | .[] | .[] | select (.status == "unconfined") | .pid' |
+          xargs --verbose --no-run-if-empty --delimiter='\n' \
+          kill
+        '';
+        commonOpts = p: "--verbose --show-cache ${lib.optionalString (!p.enforce) "--complain "}${p.profile}";
+        in {
+        Type = "oneshot";
+        RemainAfterExit = "yes";
+        ExecStartPre = "${pkgs.apparmor-utils}/bin/aa-teardown";
```

Comment:
running `aa-teardown` means to unload all AppArmor profiles, and to remove the confinement from all running processes - and the only way to re-confine them (after loading profiles again) is to restart these processes.

Is this really what you want?

Same question for `ExecStop` which will unfortunately also be used if someone runs `systemctl restart apparmor` (`restart` maps to `stop`, then `start`).

If your goal is to unload deleted profiles, have a look at `aa-remove-unknown` instead. Note that it will also remove profiles that don't exist as files in `/etc/apparmor.d/`, for example autogenerated profiles for containers.

As a sidenote - to avoid this problem, upstream `apparmor.service` has `ExecStop=/bin/true`. That's not perfect (for example, not even deleted profiles will get unloaded), but it errors out on the secure side.

(Disclaimer: I'm not familiar with NixOS or this code, and the above comment assumes that this code is used to write the `apparmor.service` systemd unit.)